### PR TITLE
[feature] [server] add dbStorage_readAheadCacheBatchBytesSize properties when read ahead entries

### PR DESCRIPTION
**Pull request by StevenLuMT**
Descriptions of the changes in this PR:

### Motivation

read ahead only limits the number of entries, which is not precise enough. so we need to add the bytes limit to make the prefetch to be more controllable

### Changes

1.  add dbStorage_readAheadCacheBatchBytesSize properties 
2. when dbStorage_readAheadCacheBatchBytesSize is setted, use it to limit the bytes for read ahead entries.

### TestCase 

1. Regression test case:  DbLedgerStorageReadCacheTest#chargeReadAheadCacheRegressionTest
2. Unit test case:   DbLedgerStorageReadCacheTest#chargeReadAheadCacheUnitTest
3. Perf test case:   DbLedgerStorageReadCacheTest#compareDiffReadAheadPerfTest

perf test results demonstrating improvements with this option:
case(read ahead cache by limit bytes size) get less cachemiss,it is suitable for large messages, reduce the pollution of readAhead large messages to readCache
image

### CODE REVIEW

**Comment by StevenLuMT** [comment]
@eolivelli @dlg99 @zymap @Shoothzj @HQebupt @AnonHxy 
needs review, thanks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java [hunk 1]
```diff
@@ -695,6 +696,18 @@ private void fillReadAheadCache(long orginalLedgerId, long firstEntryId, long fi
         }
     }
 
+    private boolean chargeReadAheadCache(int currentReadAheadCount, long currentReadAheadBytes) {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java:699]: can we add unit tests for this ? you can make it a static method, it will be very easy to write an unit test.
the logic is not straightforward, a test with edge cases will help

**Review by eolivelli — CHANGES_REQUESTED** [review]
Perfect.

I added an ask about writing an unit test

**Review by Shoothzj — APPROVED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java [hunk 2]
```diff
@@ -695,6 +696,18 @@ private void fillReadAheadCache(long orginalLedgerId, long firstEntryId, long fi
...
+        // compatible with old logic
+        boolean chargeSizeCondition = currentReadAheadCount < readAheadCacheBatchSize
+                && currentReadAheadBytes < maxReadAheadBytesSize;
+        if (readAheadCacheBatchBytesSize > 0) {
```
Comment (zymap) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java:703]: We should also check it is smaller than the max read cache size.

**Review by zymap — COMMENTED** [review]
Would it be more clearly to add a configuration to change the percentage of the readCacheMaxSize? Currently, it is 50% of readCacheMaxSize. Using a bytes' configuration, you need to calculate the size and take care of the readCacheMaxSize. If we use percentage, we don't need to consider that.

**Review by hangc0276 — CHANGES_REQUESTED** [review]
We'd better add the performance result to show this change is valuable.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java (hunk 1 above)
Comment (StevenLuMT) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java:699]: > can we add unit tests for this ? you can make it a static method, it will be very easy to write an unit test.
> the logic is not straightforward, a test with edge cases will help

Thanks for reviewing , I will add unit test today @eolivelli

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java (hunk 2 above)
Comment (StevenLuMT) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java:703]: Good catch, I will update it @zymap

**Comment by StevenLuMT** [comment]
> Do you have perf test results demonstrating improvements with this option?

Thanks for reviewing , I will add unit test today @dlg99

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java [hunk 3]
```diff
@@ -695,6 +696,18 @@ private void fillReadAheadCache(long orginalLedgerId, long firstEntryId, long fi
...
+            // exact limits limit the size and count for each batch
+            chargeSizeCondition = currentReadAheadCount < readAheadCacheBatchSize
+                    && currentReadAheadBytes < readAheadCacheBatchBytesSize;
+        }
```
Comment (AnonHxy) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java:707]: If the `readAheadCacheBatchBytesSize` must smaller than `maxReadAheadBytesSize`, maybe we could change    like this:
```suggestion
        if (chargeSizeCondition && readAheadCacheBatchBytesSize > 0) {
            // exact limits limit the size and count for each batch
            chargeSizeCondition = currentReadAheadBytes < readAheadCacheBatchBytesSize;
        }
```

File: conf/bk_server.conf [hunk 4]
```diff
@@ -749,6 +749,9 @@ gcEntryLogMetadataCacheEnabled=false
 # By default it will be allocated to 25% of the available direct memory
 # dbStorage_readAheadCacheMaxSizeMb=
 
+# How many entries' bytes to pre-fill in cache after a read cache miss
+# dbStorage_readAheadCacheBatchBytesSize=-1
```
Comment (AnonHxy) [inline conf/bk_server.conf:753]: How about change the default value to 0?  Because 0 will also disable this feature.

And we'd better add a comments like this:
`Default is 0. 0 or less disables this feature`

File: conf/bk_server.conf [hunk 5]
```diff
@@ -749,6 +749,9 @@ gcEntryLogMetadataCacheEnabled=false
 # By default it will be allocated to 25% of the available direct memory
 # dbStorage_readAheadCacheMaxSizeMb=
 
+# How many entries' bytes to pre-fill in cache after a read cache miss,default is 0 or less disables this feature
```
Comment (AnonHxy) [inline conf/bk_server.conf:752]:
```suggestion
# How many entries' bytes to pre-fill in cache after a read cache miss. Default is -1. 0 or less disables this feature
```

File: conf/bk_server.conf (hunk 4 above)
Comment (StevenLuMT) [reply conf/bk_server.conf:753]: > How about change the default value to 0? Because 0 will also disable this feature.
> 
> And we'd better add a comments like this: `Default is 0. 0 or less disables this feature`

updated,thanks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java (hunk 3 above)
Comment (StevenLuMT) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java]: > If the `readAheadCacheBatchBytesSize` must smaller than `maxReadAheadBytesSize`, maybe we could change like this:

The logic is not inherited, it cannot be changed like this, thanks

**Comment by StevenLuMT** [comment]
I have add a unit test(TestBookieImpl) to cover this function,and add the performance result to show this change is valuable.
help me review,thanks @dlg99 @eolivelli @hangc0276 @zymap

**Review by AnonHxy — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
Lgtm

**Comment by StevenLuMT** [comment]
> perf test

@dlg99 @hangc0276 add three testcase: 
1. Regression test case: DbLedgerStorageReadCacheTest#chargeReadAheadCacheRegressionTest
2. Unit test case: DbLedgerStorageReadCacheTest#chargeReadAheadCacheUnitTest
3. Perf test case: DbLedgerStorageReadCacheTest#compareDiffReadAheadPerfTest

**Comment by StevenLuMT** [comment]
rerun failure checks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java (hunk 1 above)
Comment (StevenLuMT) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java]: > I saw the unit tests, very good.
> 
> But here I meant to add a "unit test" that tests this method (you can make it "static", no need to start a Bookie)

updated,thanks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java [hunk 6]
```diff
@@ -695,6 +696,19 @@ private void fillReadAheadCache(long orginalLedgerId, long firstEntryId, long fi
         }
     }
 
+    protected boolean chargeReadAheadCache(int currentReadAheadCount, long currentReadAheadBytes) {
+        // compatible with old logic
+        boolean chargeSizeCondition = currentReadAheadCount < readAheadCacheBatchSize
+                && currentReadAheadBytes < maxReadAheadBytesSize;
+        if (readAheadCacheBatchBytesSize > 0) {
+            // exact limits limit the size and count for each batch
+            chargeSizeCondition = currentReadAheadCount < readAheadCacheBatchSize
```
Comment (hangc0276) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java:705]: if (readAheadCacheBatchBytesSize > 0 && chargeSizeCondition) {
    chargeSizeCondition = chargeSizeCondition && currentReadAheadBytes < readCacheMaxSize;
}

**Review by hangc0276 — APPROVED** [review]
Good job!
